README.md
# Support

If you like the "ZX NEX File Viewer" please consider supporting it.

<a href="https://github.com/sponsors/maziac" title="Github sponsor">
	<img src="assets/remote/button_donate_sp.png" />
</a>
&nbsp;&nbsp;
<a href="https://www.paypal.com/donate/?hosted_button_id=K6NNLZCTN3UV4&locale.x=en_DE&Z3JncnB0=" title="PayPal">
	<img src="assets/remote/button_donate_pp.png" />
</a>


# ZX NEX File Viewer


This extension adds a readonly file viewer for ZX Spectrum NEX files.
.nex is a special file format for loading ZX Next programs.
The viewer shows registers, a memory dump and the included screen images.

The NEX file format specification can be found [here](https://wiki.specnext.dev/NEX_file_format).


## Installation

Install through Visual Studio Code Marketplace.
The extension is called "ZX NEX File Viewer".


# Usage

Just click on a file with the *.nex extension.
The file is opened and shown in the text editor area.

![](assets/remote/usage.gif)

Alternatively you can open a NEX file via "Open With...":

![](assets/remote/usage-open-with.gif)

When hovering over a value you can get additional info like the offset or the value in decimal:

![](assets/remote/hovering.gif)


CHANGELOG.md
# Changelog

## 1.4.0
- Changed to esbuild. Size decreased.

## 1.3.0
- Contents remains even if editor is hidden.

## 1.2.0
- Supporting vscode workspace trust.

## 1.1.2
- Buttons updated.

## 1.1.1
- vscode category changed.

## 1.1.0
- NEX file viewer will update automatically if file changes in background.
- Load screen now opened by default.

## 1.0.0
- First beta test version.
- Register view.
- Memory dumps.
- Screen image.

## 0.1.0
- Initial version.


src/extension.ts
import * as vscode from 'vscode';
import {EditorProvider} from './editorprovider';


export function activate(context: vscode.ExtensionContext) {
    // Enable logging.
    configure(context);

    // Check for every change.
    context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(event => {
        configure(context, event);
    }));

    // Register custom readonly editor provider
    const viewProvider = new EditorProvider();
    vscode.window.registerCustomEditorProvider('nex-fileviewer.viewer', viewProvider, {webviewOptions: {enableFindWidget: true, retainContextWhenHidden: true}});
}


/**
 * Reads the configuration.
 */
function configure(context: vscode.ExtensionContext, event?) {
    /*
    const settings = vscode.workspace.getConfiguration('z80-instruction-set');

    // Note: don't add 'language' property, otherwise other extension with similar file pattern may not work.
    // If the identifier is missing it also don't help to define it in package.json. And if "id" would be used it clashes again with other extensions.
    const asmFiles: vscode.DocumentSelector = { scheme: "file", pattern: settings.files};

     // Enable/disable hovering
    if(settings.enableHovering) {
        if(!regHoverProvider) {
            // Register
            regHoverProvider = vscode.languages.registerHoverProvider(asmFiles, new HoverProvider());
            context.subscriptions.push(regHoverProvider);
        }
    }
    else {
        if(regHoverProvider) {
            // Deregister
            regHoverProvider.dispose();
            regHoverProvider = undefined;
        }
    }
    */
}



// this method is called when your extension is deactivated
export function deactivate() {
}


src/html/parser.ts
import {vscode} from "./vscode-import";



/**
 * This js script parses a file, does all the decoding and presents the
 * data in the webview.
 * It is done as script inside the webview (opposed to creating a html file in
 * the extension) to allow lazy loading.
 * Large blocks of data are skipped in the initial pass and decoded only
 * when needed. I.e. when the user expands an item.
 */


// The data to parse.
export let dataBuffer: number[];

// Index into snaData
export let lastOffset: number;

// The last retrieved data size.
export let lastSize: number;

// The root node for parsing. New objects are appended here.
export let lastNode: any;

// The correspondent node for the details.
let lastContentNode: any;

// The last node used for the title.
let lastNameNode: any;

// The last node used for the value.
let lastValueNode: any;

// The last node used for the short description.
export let lastDescriptionNode: any;

// The last node used for the long description.
let lastLongDescriptionNode: any;



/**
 * Call to check a value.
 * Does nothing.
 * You can set a breakpoint here.
 */
export function assert(condition: boolean) {
	if (!condition) {
		console.log("Error!");
	}
}


/**
 * Convert array to base 64 string.
 */
export function arrayBufferToBase64(buffer) {
	let binary = '';
	const bytes = [].slice.call(new Uint8Array(buffer));
	bytes.forEach((b) => binary += String.fromCharCode(b));
	return window.btoa(binary);
}


/**
 * Creates a node and appends it to lastNode.
 * @param name The name of the value. E.g. "SP".
 * @param valString The value to show.
 * @param shortDescription A short description of the entry.
 */
export function createNode(name: string, valString = '', shortDescription = ''): HTMLDetailsElement {
	// Create new node
	const node = document.createElement("DETAILS") as HTMLDetailsElement;
	node.classList.add("nomarker");
	//node.classList.add("basenode");
	const lastOffsetHex = getHexString(lastOffset, 4);
	const lastSizeHex = getHexString(lastSize, 4);
	const html = `
<summary>
	<div class="offset" title="Offset\nHex: ${lastOffsetHex}">${lastOffset}</div>
	<div class="size" title="Size\nHex: ${lastSizeHex}">${lastSize}</div>
	<div class="name">${name}</div>
	<div class="value">${valString}</div>
	<div class="description">${shortDescription}</div>
</summary>
<div class="indent"></div>
`;
	node.innerHTML = html;

	// Get child objects
	const childrenNode = node.childNodes;
	lastContentNode = childrenNode[3];
	const summary = childrenNode[1];
	const children = summary.childNodes;
	lastNameNode = children[5];
	lastValueNode = children[7];
	lastDescriptionNode = children[9];
	const descriptionChildren = lastDescriptionNode.childNodes;
	lastLongDescriptionNode = descriptionChildren[3];

	// Append it
	lastNode.appendChild(node);

	// Return
	return node;
}


/**
 * Adds a long description.
 * Will be shown when expanded.
 */
export function addDescription(longDescription: string) {
	//lastLongDescriptionNode.innerHTML = longDescription;
	beginDetails();
	createDescription(convertLineBreaks(longDescription));
	endDetails();
}


/**
 * Converts \n into <br>.
 */
function convertLineBreaks(s: string) {
	return s.replace(/\n/g, '<br>');
}


/**
 * Sets lastNode to it's last child.
 * This begins a details sections.
 * Indents.
 */
function beginDetails() {
	const node = lastNode.lastChild;
	node.classList.remove("nomarker");
	lastNode = lastContentNode;
}


/**
 * Ends a details sections.
 * Sets lastNode to it's parent.
 * Indents.
 */
function endDetails() {
	lastNode = lastNode.parentNode.parentNode;
}


/**
 * Parses the details of an object.
 * Parsing is done immediately.
 * Uses begin/endDetails.
 * @param func The function to call to parse/decode the data.
 */
export function addDetailsParsing(func: () => void) {
	// "Indent"
	beginDetails();
	// Call function
	const bakLastOffset = lastOffset;
	const bakLastSize = lastSize;
	lastSize = 0;
	func();
	lastOffset = bakLastOffset;
	lastSize = bakLastSize;
	// Close/leave
	endDetails();
}


/**
 * Installs a listener for the toggle event.
 * The parsing of the data is delayed until toggling.
 * @param func The function to call to parse/decode the data.
 */
export function addDelayedDetailsParsing(func: () => void) {
	// Get nodex
	const detailsNode = lastNode.lastChild;
	detailsNode.classList.remove("nomarker");
	// Attach attribute
	detailsNode.setAttribute('data-index', lastOffset.toString());
	// Install listener
	if (func) {
		detailsNode.addEventListener("toggle", function handler(this: any, event: any) {
			// Get parse node and index
			lastNode = event.target;
			const indexString = lastNode.getAttribute('data-index');
			lastOffset = parseInt(indexString);
			lastSize = 0;
			func();
			this.removeEventListener("toggle", handler);
		});
	}
}


/**
 * Creates a description line of contents.
 * Is gray.
 * @param descr The description string. Any linebreaks are converted into '<br>'.
 */
export function createDescription(descr: string) {
	// Create new node
	const node = document.createElement("DIV");
	// Add description
	node.innerHTML = convertLineBreaks(descr);
	// Apply style gray
	node.classList.add('gray');
	// Append it
	lastNode.appendChild(node);
}


/**
 * Creates an element with a number of columns.
 * Without any index and size columns.
 * @param name The name of the value. E.g. "SP".
 * @param valString The value to show.
 * @param shortDescription A short description of the entry.
 */
export function createSimpleRow(name: string, valString = '', shortDescription = '') {
	// Create new node
	const node = document.createElement("DETAILS");
	node.classList.add("nomarker");
	const html = `
<summary>
	<div class="offset"></div>
	<div class="size"></div>
	<div class="name">${name}</div>
	<div class="value">${valString}</div>
	<div class="description">${shortDescription}</div>
</summary>
<div class="indent"></div>
`;
	node.innerHTML = html;

	// Get child objects
	const childrenNode = node.childNodes;
	lastContentNode = childrenNode[3];
	const summary = childrenNode[1];
	const children = summary.childNodes;
	lastNameNode = children[5];
	lastValueNode = children[7];
	lastDescriptionNode = children[9];
	const descriptionChildren = lastDescriptionNode.childNodes;
	lastLongDescriptionNode = descriptionChildren[3];

	// Append it
	lastNode.appendChild(node);

	// Return
	return node;
}


/**
 * Adds a hover text to lastValueNode.
 * @param hoverValueString String to show on hover for the title. Can be undefined.
 */
export function addHoverValue(hoverValueString: string) {
	lastValueNode.title = hoverValueString;
}


/**
 * Returns a hex string.
 * @param value The value to convert.
 * @param size The number of digits (e.g. 2 or 4)
 * @returns E.g. "0Fh" or "12FAh"
 */
export function getHexString(value: number, size: number): string {
	if (value == undefined)
		return "".padStart(size, '?');
	const s = value.toString(16).toUpperCase().padStart(size, '0');
	return s;
}


/**
 * Advances the offset (from previous call) and
 * stores the size for reading.
 * @param size The number of bytes to read.
 */
export function read(size: number) {
	lastOffset += lastSize;
	lastSize = size;
}


/**
 * Sets lastSize directly (for special cases).
 * @param size The number of bytes to read.
 */
export function setLastSize(size: number) {
	lastSize = size;
}


/**
 * Reads the value from the buffer.
 */
export function getValue(): number {
	let value = dataBuffer[lastOffset];
	let factor = 1;
	for (let i = 1; i < lastSize; i++) {
		factor *= 256;
		value += factor * dataBuffer[lastOffset + i];
	}
	return value;
}


/**
 * @returns The value from the dataBuffer as decimal string.
 */
export function decimalValue(): string {
	const val = getValue();
	return val.toString();
}


/**
 * @returns The value from the dataBuffer as hex string.
 */
export function hexValue(): string {
	const val = getValue();
	let s = val.toString(16).toUpperCase();
	s = s.padStart(lastSize * 2, '0');
	return s;
}

/**
 * @returns The value from the dataBuffer as hex string + "0x" in front.
 */
export function hex0xValue(): string {
	return '0x'+hexValue();
}


/**
 * @param bit The bit to test
 * @returns The bit value (0 or 1) from the dataBuffer as string.
 */
export function bitValue(bit: number): string {
	const val = getValue();
	const result = (val & (1 << bit)) ? '1' : '0';
	return result;
}


/**
 * Converts a value into a bit string.
 * @param value The value to convert.
 * @param size The size of the value, e.g. 1 byte o r 2 bytes.
 * @returns The value from the dataBuffer as bit string. e.g. "0011_0101"
 */
export function convertBitsToString(value: number, size: number): string {
	let s = value.toString(2);
	s = s.padStart(size * 8, '0');
	s = s.replace(/.{4}/g, '$&_');
	// Remove last '_'
	s = s.substring(0, s.length - 1);
	return s;
}

/**
 * @returns The value from the dataBuffer as bit string. e.g. "0011_0101"
 */
export function bitsValue(): string {
	const val = getValue();
	return convertBitsToString(val, lastSize);
}

/**
 * Reads a text of given size.
 * @returns The data as string.
 */
export function stringValue(): string {
	let s = '';
	for (let i = 0; i < lastSize; i++) {
		const c = dataBuffer[lastOffset + i];
		s += String.fromCharCode(c);
	}
	return s;
}


/**
 * Is called if the user opens the details of an item.
 * Decodes the data.
 * @param displayOffset The displayOffset is added to the index before displaying.
 * @param hoverRelativeOffset You can replace the default relative Offset hover text.
 */
export function createMemDump(displayOffset = 0, hoverRelativeOffset = 'Relative Offset') {
	let html = '';
	let prevClose = '';

	// In case of an error, show at least what has been parsed so far.
	try {
		// Loop given size
		for (let i = 0; i < lastSize; i++) {
			const k = i % 16;
			// Get value
			const iOffset = lastOffset + i;	// For indexing
			const iRelOffset = displayOffset + i;	// For display
			const val = dataBuffer[iOffset];
			const valString = getHexString(val, 2);
			const valIntString = val.toString();
			const iRelOffsetHex = getHexString(iRelOffset, 4);

			// Start of row?
			if (k == 0) {
				// Close previous
				html += prevClose;
				prevClose = '</div>';

				// Check for same values
				let l = i + 1
				for (; l < lastSize; l++) {
					if (val != dataBuffer[lastOffset + l])
						break;
				}
				const l16 = l - (l % 16);
				if (l16 > i + 16) {
					// At least 2 complete rows contains same values
					if (l == lastSize)
						i = lastSize - 1;	// end
					else
						i = l16 - 1;
					const iRelOffsetHexEnd = getHexString(displayOffset + i, 4);

					const hoverTextOffset = 'Offset (hex): ' + getHexString(iOffset, 4) + '-' + getHexString(lastOffset + i, 4) + '\nValue (dec): ' + valIntString;

					const hoverTextRelOffset = 'Relative offset (dec): ' + iRelOffset + '-' + (displayOffset + i) + '\nValue (dec): ' + valIntString;

					html += '<div>';
					html += '<span class="indent mem_offset" title="' + hoverTextOffset + '">' + iOffset + '-' + (lastOffset + i) + '</span>';
					html += '<span class="mem_offset" title="' + hoverTextRelOffset +'"> (0x' + iRelOffsetHex + '-0x' + iRelOffsetHexEnd + '): </span>';
					html += '<span title="Value (dec): '+valIntString+'"> contain all ' + valString + '</span>';
					continue;
				}

				// Afterwards proceed normal
				const iOffsetHex = getHexString(iOffset, 4);
				html += `<div class="mem_dump">
					<div class="indent mem_offset" title = "Offset\nHex: ${iOffsetHex}">${iOffset}</div>
				<div class="mem_rel_offset" title="${hoverRelativeOffset}\nDec: ${iRelOffset}">(0x${iRelOffsetHex})</div>
				`;
			}

			// Convert to html
			const hoverText = 'Offset (hex): ' + getHexString(iOffset, 4) + '\nOffset (dec): ' + iOffset + '\nRelative offset (hex): ' + iRelOffsetHex + '\nRelative offset (dec): ' + iRelOffset + '\nValue (dec): ' + valIntString;
			html += '<div class="mem_byte" title="' + hoverText + '">' + valString + '&nbsp;</div>';
		}
		// Close
		html += prevClose;
	}
	catch (e) {
		// Close
		html += prevClose;
		// Error while parsing
		html += '<div class="error indent">Error while parsing.</div>';
	}

	// Append
	lastNode.innerHTML += html;
}


/**
 * Starts the parsing.
 */
export function parseInit(data: number[]) {
	// Reset
	dataBuffer = data;
	lastOffset = 0;
	lastSize = 0;
	lastNode = document.getElementById("div_root");
}


/**
 * Copies the complete html of the document to the clipboard.
 */
export function copyHtmlToClipboard() {
	const copyText = document.documentElement.innerHTML;
	navigator.clipboard.writeText(copyText);
}


// At the end send a message to indicate that the webview is ready to receive
// data.
vscode.postMessage({
	command: 'ready'
});


src/html/main.ts
import {assert, getValue, read, createNode, addDetailsParsing, hex0xValue, addHoverValue, addDelayedDetailsParsing, createMemDump, lastOffset, parseInit, dataBuffer, lastNode, stringValue, createDescription, lastSize, lastDescriptionNode, addDescription, decimalValue, bitsValue, bitValue, hexValue, setLastSize} from "./parser";
import {banksValue, coreVersionValue, createCopperDump, createLayer2Screen, createLayer2Screen320, createLayer2Screen640, createLoResScreen, createPalette, createPaletteImage, createPaletteWithOffset, createTimexHiColScreen, createTimexHiResScreen, getPalette, getZxNextDefaultPalette} from "./zxnext";
import {createUlaScreen, getMemBankPermutation, zxColorValue} from './zxspecific';



/**
 * The main program which starts the decoding.
 * Must implement the parseRoot function.
 */



//---- Parse the data (root level) --------
function parseRoot() {
	// Meta info
	let html = '<div>ZX NEX File View.</div>';
	const length = dataBuffer.length;
	html += '<div><b>Length:</b> ' + length + '</div>';
	html += '<br>';
	lastNode.innerHTML = html;
	// End of meta info


	try {
		// Header
		let hasCopperCode;
		let loadScreens;
		let loadScreens2;
		let hiresColor;
		read(512);
		createNode('Header').open = true;
		addDetailsParsing(() => {
			read(4);
			const nextString = stringValue();
			createNode('NEXT', nextString);
			if (nextString != 'Next') {
				// Not a NEX file
				lastDescriptionNode.classList.add('error');
				lastDescriptionNode.textContent = 'Not a NEX file.';
			}

			read(4);
			const version = stringValue();
			let unOfficial;
			if (version == 'V1.3')
				unOfficial = '(unofficial)';
			createNode('VERSION', version, unOfficial);

			read(1);
			const ramRequired = getValue();
			let ramReqValue;
			switch (ramRequired) {
				case 0: ramReqValue = '768k'; break;
				case 1: ramReqValue = '1792k'; break;
				default: ramReqValue = ramRequired.toString();
			}
			createNode('RAM_REQUIRED', ramReqValue, 'The required RAM.');
			addDescription('0 = 768k, 1 = 1792k');
			addHoverValue('Dec: ' + decimalValue());

			read(1);
			createNode('NUM_BANKS', decimalValue(), 'Number of 16k Banks to Load: 0-112');

			read(1);
			loadScreens = getValue();
			createNode('LOAD_SCREENS', bitsValue(), 'Loading-screen blocks in file');
			addDetailsParsing(() => {
				createDescription(`
128 = no palette block, 64 = "flags 2" in V1.3 part of header define screen, 16 = Hi-Colour, 8 = Hi-Res, 4 = Lo-Res, 2 = ULA, 1 = Layer2

The loader does use common banks to load the graphics into, and show it from, i.e. bank5 for all ULA related modes and banks 9,10 and 11 for Layer2 graphics (loading these banks afterwards as regular bank will thus replace the shown data on screen).

Only Layer2, Tilemap and Lo-Res screens expect the palette block (unless +128 flag set). While one can include multiple screen data in single file (setting up all relevant bits), the recommended/expected usage is to have only one type of screen in NEX file.`);
				createNode('No palette block', bitValue(7), 'Bit 7');
				createNode('flags 2', bitValue(6), 'Bit 6');
				createNode('Unused', bitValue(7), 'Bit 5');
				createNode('Hi-Colour', bitValue(4), 'Bit 4');
				createNode('Hi-Res', bitValue(3), 'Bit 3');
				createNode('Lo-Res', bitValue(2), 'Bit 2');
				createNode('ULA', bitValue(1), 'Bit 1');
				createNode('Layer 2', bitValue(0), 'Bit 0');
			});

			read(1);
			createNode('BORDER_COLOR', zxColorValue(), 'Border Color: 0-7');
			addHoverValue('Dec: ' + decimalValue());

			read(2);
			createNode('SP', hex0xValue(), 'Stack pointer');
			addHoverValue('Dec: ' + decimalValue());

			read(2);
			createNode('PC', hex0xValue(), 'Program counter');
			addDescription("0 = don't run, just load");
			addHoverValue('Dec: ' + decimalValue());

			read(2);
			createNode('NUM_FILES', decimalValue(), 'Number of extra files');
			addDescription("Obsolete");

			read(112);
			createNode('BANKS', banksValue(), 'Array of included banks');
			addDetailsParsing(() => {
				createDescription('byte flag (0/1) of 16k banks included in the file - this array is in regular order 0..111, i.e. bank5 in file will set 1 to header byte at offset 18+5 = 23, but the 16kiB of data for bank 5 are first in the file (order of bank data in file is: 5,2,0,1,3,4,6,7,8,9,10,...,111)');
				setLastSize(0);
				for (let i = 0; i < 112; i++) {
					read(1);
					createNode('Bank' + i, decimalValue());
				}
			});

			read(1);
			createNode('LOAD_BAR', decimalValue(), 'Layer2 "loading bar"');
			addDescription('0 = OFF, 1 = ON(works only in combination with Layer2 screen data) ');

			read(1);
			createNode('LOAD_COLOR', decimalValue(), '"Loading bar" color');
			addDescription('"Loading bar" color (0..255) (for 640x256x4 mode the byte defines pixels pair)');

			read(1);
			createNode('LOAD_DELAY', decimalValue(), 'Loading delay per bank');
			addDescription('Loading delay per bank (0..255 amount of frames), 0 = no delay');

			read(1);
			createNode('START_DELAY', decimalValue(), 'Start delay');
			addDescription('Start delay (0..255 amount of frames), 0 = no delay');

			read(1);
			createNode('PRESERVE_NEXT_REGS', decimalValue(), 'Preserve Next-Registers');
			addDescription('Preserve current Next-Registers values (0 = reset machine state, 1 = preserve)');

			read(3);
			createNode('CORE_VERSION', coreVersionValue(), 'Required core version');
			addDetailsParsing(() => {
				createDescription('Required core version, three bytes 0..15 "major", 0..15 "minor", 0..255 "subminor" version numbers. (core version is checked only when reported machine-ID is 10 = "Next", on other machine or emulator=8 the latest loaders will skip the check)');
				//lastSize = 0;
				read(1);
				createNode('MAJOR', decimalValue());
				read(1);
				createNode('MINOR', decimalValue());
				read(1);
				createNode('SUB_MINOR', decimalValue());
			});


			read(1);
			hiresColor = getValue();
			createNode('HIRES_COLOR', bitsValue(), 'Timex HiRes 512x192 mode color');
			addDescription(`Timex HiRes 512x192 mode color, encoded as for port 255 = bits 5-3. I.e. values 0, 8, 16, .., 56 (0..7 * 8)
When screens 320x256x8 or 640x256x4 are used, this byte is re-used as palette offset for Layer 2 mode, values 0..15`);

			read(1);
			createNode('ENTRY_BANK', decimalValue(), 'Bank to be mapped into slot 3');
			addDescription('Entry bank = bank to be mapped into slot 3 (0xC000..0xFFFF address space), the "Program Counter" (header offset +14) and "File handle address" (header offset +140) are used by NEX loader after the mapping is set (The default ZX128 has bank 0 mapped after reset, which makes zero value nice default).');

			read(2);
			createNode('FILE_HANDLE_ADDR', hexValue(), 'File handle address');
			addDescription('0 = NEX file is closed by the loader, 1..0x3FFF values (1 recommended) = NEX loader keeps NEX file open and does pass the file handle in BC register, 0x4000..0xFFFF values (for 0xC000..0xFFFF see also "Entry bank") = NEX loader keeps NEX file open and the file handle is written into memory at the desired address.');
			addHoverValue('Dec: ' + decimalValue());
			// End of version 1.2 header
			assert(lastOffset + lastSize == 142);

			if (version <= "V1.2") {
				// Version 1.0 or 1.1
				const remaining = 512 - 142;
				read(remaining);
				createNode('RESERVED');
				addDescription('Reserved for future extensions.');
				addDelayedDetailsParsing(() => {
					read(remaining);
					createMemDump();
				});
			}
			else {
				// Version 1.2 or bigger
				read(1);
				createNode('EXPBUS_ENABLE', decimalValue(), 'Enable Expansion Bus');
				addDescription('0 = disable Expansion Bus by setting top four bits of Expansion Bus Enable Register ($80) to 0, 1 = do nothing (does apply only to cores 3.0.5+)');

				read(1);
				createNode('HAS_CHECKSUM', decimalValue());
				addDescription('1 = Has checksum value, checksum algorithm is CRC-32C (Castagnoli), value itself is at the very end of the header block');

				read(4);
				createNode('BANKS_OFFSET', hexValue(), 'File offset of first bank data');
				addDescription('File offset of first bank data (when loader is parsing known version, it should know where the banks start without this value, but it may use it for extra check whether the parsing of optional blocks between header and first bank was done correctly for files V1.3+, or it may even try to partially-load unknown future versions of NEX files by skipping unknown blocks between header and banks data, although that may lead to unexpected state for the app)');
				addHoverValue('Dec: ' + decimalValue());

				read(2);
				createNode('CLI_BUFFER_ADDR', decimalValue(), 'CLI buffer address');
				addDescription('CLI buffer address (after "Entry bank" is paged in), 0 = no buffer');

				read(2);
				createNode('CLI_BUFFER_SIZE', decimalValue(), 'CLI buffer size');
				addDescription('When address and size are provided, the original argument line passed to NEX loader will be copied to defined buffer (and truncated to "size", shorter string may be zero/colon/enter terminated as any other BASIC line) and register DE is set to the buffer address. The maximum size is 2048 bytes (longer lines can be probably salvaged from Bank 5 memory if the NEX file is not loading that bank and the app code search for the original line on its own).');

				read(1);
				loadScreens2 = getValue();
				createNode('LOAD_SCREENS_2', decimalValue(), 'Loading screen flags 2 ');
				addDescription(`When first flag has bit6 +64 set (+128 no-palette is valid for new cases too, other old bits should be NOT mixed with new modes):
1 = Layer 2 320x256x8bpp, blocks: [512B palette +] 81920B data
2 = Layer 2 640x256x4bpp, blocks: [512B palette +] 81920B data
 (HiRes color value 0..15 is used as L2 palette offset - does apply to two new Layer 2 modes)
 For Layer 2 banks 9,10,11,12,13 are used to display the loading screen.
3 = Tilemode screen, block: [512B palette] (plus four configuration bytes in header at offset 154)
 Tilemap data are stored in regular (!) bank 5 - no specialized data block is used.`);

				read(1);
				hasCopperCode = getValue();
				createNode('HAS_COPPER_CODE', decimalValue(), 'Inclusion of copper code');
				addDescription('1 = Has copper code block, extra 2048B block after last screen data block, which will be set to Copper and the copper will be started with %01 control code (reset CPC to 0, and start). This can be used for example as "loading screen animation" feature (be aware the timing of load may vary greatly).');

				read(4);
				createNode('TILE_SCR_CONFIG', decimalValue(), 'Tilemode screen configuration');
				addDescription('When Tilemode screen: four bytes array, values to set NextRegs: $6B, $6C, $6E, $6F');

				read(1);
				createNode('BIG_L2_BAR_POSY', decimalValue(), 'Loading bar y-position');
				addDescription('When Layer 2 320x256 or 640x256 screen + loading bar: loading bar Y position (bar is 2px tall, so 254 is very bottom)');

				read(349);
				createNode('RESERVED');
				addDescription('Reserved for future extensions, set to zero in older versions of file format.');
				addDelayedDetailsParsing(() => {
					read(349);
					createMemDump();
				});

				read(4);
				createNode('CRC32C', decimalValue(), 'Optional checksum');
				addDescription('Optional CRC-32C (Castagnoli) - is calculated by checksumming (initial value is zero) file content from offset 512 (after this field) till end of file, including the optional custom binary appended after regular NEX file. And then continuing with checksumming the first 508 bytes of the header (stopping ahead of this field). This is intended for PC tools working with NEX files, or extra check tool running on Next, but not for regular loading of NEX files (value is stored as "uint32_t" in little-endian way).');
			}

			// Check size
			assert(lastOffset == 512);
		});


		// Palette
		let palette;
		if (!(loadScreens & 0b1000_0000)	// Has no-palette not set
			&& (loadScreens & 0b0100_0101))	// Layer2, LORES or LOADSCR2 is set
		{
			read(512);
			palette = getPalette();
			createNode('Palette');
			addDetailsParsing(() => {
				createDescription('Optional palette (for Layer2, LoRes or Tilemap screen).\nPalette consists of 512 bytes (256 palette items from index 0), in 9b colour format: first byte is RRRG_GGBB, second byte is P000_000B (P is priority flag for Layer 2 colours).');
				read(512);
				createPaletteImage();
				createNode('As index array');
				addDelayedDetailsParsing(() => {
					read(512);
					createPalette();
				});
			});
		}
		else {
			// Use default palette
			palette = getZxNextDefaultPalette();
		}

		// Layer 2 loading screen
		if (loadScreens & 0b0000_0001) {
			read(49152);
			createNode('LAYER2_LOAD_SCREEN', '', 'Layer 2 loading screen').open = true;
			addDetailsParsing(() => {
				createDescription('Palette consists of 512 bytes (256 palette items from index 0), in 9b color format: first byte is RRRG_GGBB, second byte is P000_000B (P is priority flag for Layer 2 colours).');
				read(49152);
				createLayer2Screen(palette);
				createNode('Memory dump');
				addDelayedDetailsParsing(() => {
					read(49152);
					createMemDump();
				});
			});
		}

		// ULA loading screen
		if (loadScreens & 0b0000_0010) {
			read(6912);
			createNode('ULA_LOAD_SCREEN', '', 'Classic ULA loading screen').open = true;
			addDetailsParsing(() => {
				read(6912);
				createUlaScreen();
				createNode('Memory dump');
				addDelayedDetailsParsing(() => {
					read(6912);
					createMemDump();
				});
			});
		}

		// LoRes loading screen
		if (loadScreens & 0b0000_0100) {
			read(12288);
			createNode('LORES_LOAD_SCREEN', '', 'LoRes (128x96) loading screen').open = true;
			addDetailsParsing(() => {
				createDescription('Palette consists of 512 bytes (256 palette items from index 0), in 9b color format: first byte is RRRG_GGBB, second byte is P000_000B (P is is not used for LoRes).');
				read(12288);
				createLoResScreen(palette);
				createNode('Memory dump');
				addDelayedDetailsParsing(() => {
					read(12288);
					createMemDump();
				});
			});
		}

		// Timex HiRes loading screen
		if (loadScreens & 0b0000_1000) {
			const timexInkColor = (hiresColor >> 2) & 0b111;
			read(12288);
			createNode('HIRES_LOAD_SCREEN', '', 'Timex HiRes (512x192) loading screen').open = true;
			addDetailsParsing(() => {
				createDescription(`The Timex HiRes mode supports only 2 colors.`);
				read(12288);
				createTimexHiResScreen(timexInkColor);
				createNode('Memory dump');
				addDelayedDetailsParsing(() => {
					read(12288);
					createMemDump();
				});
			});
		}

		// Timex HiCol loading screen
		if (loadScreens & 0b0001_0000) {
			read(12288);
			createNode('HICOL_LOAD_SCREEN', '', 'Timex HiCol (8x1) loading screen').open = true;
			addDetailsParsing(() => {
				createDescription(`HiCol modes uses more attributes. I.e. the attribute applies to 1 byte only.`);
				read(12288);
				createTimexHiColScreen();
				createNode('Memory dump');
				addDelayedDetailsParsing(() => {
					read(12288);
					createMemDump();
				});
			});
		}

		// Layer 2 320x256x8 or 640x256x4 loading screen
		if (loadScreens & 0b0100_0000) {
			read(81920);
			// Check which screen
			switch (loadScreens2) {
				case 1:
					// Layer 2 320x256x8bpp, blocks: [512B palette +] 81920B data
					createNode('LAYER2_320_LOAD_SCREEN', '', 'Layer 2 320x256x8bpp, 1 byte per pixel').open = true;
					addDetailsParsing(() => {
						read(81920);
						const palOffset = hiresColor & 0x0F;
						const offsPalette = createPaletteWithOffset(palette, palOffset);
						createLayer2Screen320(offsPalette);
						createNode('Memory dump');
						addDelayedDetailsParsing(() => {
							read(81920);
							createMemDump();
						});
					});
					break;
				case 2:
					// Layer 2 640x256x8bpp, blocks: [512B palette +] 81920B data
					createNode('LAYER2_320_LOAD_SCREEN', '', 'Layer 2 640x256x4bpp, 4 bits per pixel').open = true;
					addDetailsParsing(() => {
						read(81920);
						const palOffset = hiresColor & 0x0F;
						const offsPalette = createPaletteWithOffset(palette, palOffset);
						createLayer2Screen640(offsPalette);
						createNode('Memory dump');
						addDelayedDetailsParsing(() => {
							read(81920);
							createMemDump();
						});
					});
					break;
			}
		}

		// Copper code
		if (hasCopperCode == 1) {
			read(2048);
			createNode('COPPER_CODE', '', 'Copper code for copper loading-effects');
			addDelayedDetailsParsing(() => {
				read(2048);
				createCopperDump();
			});
		}

		// Memory banks
		for (let b = 0; b < 112; b++) {
			// Check byte flag
			const bank = getMemBankPermutation(b);
			const bankEnabled = (dataBuffer[18 + bank] == 1);
			if (!bankEnabled)
				continue;

			// Decode memory bank
			read(16384);
			createNode('BANK' + bank, '', '16k memory bank');
			addDelayedDetailsParsing(() => {
				read(16384);
				createMemDump();
			});
		}

		// Optional data
		const remainingLength = dataBuffer.length - lastOffset - lastSize;
		assert(remainingLength >= 0);
		if(remainingLength>0) {
			read(remainingLength);
			createNode('OPTIONAL_DATA', '', 'Optional binary data appended after "regular" NEX file content - custom format');
			addDelayedDetailsParsing(() => {
				read(remainingLength);
				createMemDump();
			});
		}

	}
	catch (e) {
		// In case of an error show the rest as memdump.
		// Partly decoded data might be re-dumped.
		const remainingLength = dataBuffer.length - lastOffset;
		assert(remainingLength >= 0);
		if (remainingLength > 0) {
			read(remainingLength);
			createNode('Remaining', '', 'Remaining data after parsing error.');
			addDelayedDetailsParsing(() => {
				read(remainingLength);
				createMemDump();
			});
		}
	}
}


//---- Handle messages from vscode extension --------
window.addEventListener('message', event => {	// NOSONAR
	const message = event.data;

	switch (message.command) {
		case 'setData':
			{
				// Parse
				parseInit(message.snaData);
				// Parse
				parseRoot();
			} break;
	}
});
